File: MatlabOctaveCodeSnippets/snip_signals_time_delay.m
Fs=8000; %sampling frequency
Ts=1/Fs; %sampling interval
N=1.5*Fs; %1.5 seconds
t=[0:N-1]*Ts;
if 1
    x = rand(1,N)-0.5; %zero mean uniformly distributed
else
    x = cos(2*pi*100*t); %cosine
end
delayInSamples=2000;
timeDelay = delayInSamples*Ts %delay in seconds
y=[zeros(1,delayInSamples) x(1:end-delayInSamples)];
SNRdb=10; %specified SNR
signalPower=mean(x.^2);
noisePower=signalPower/(10^(SNRdb/10));
noise=sqrt(noisePower)*randn(size(y));
y=y+noise;
subplot(211); plot(t,x,t,y);
[c,lags]=xcorr(x,y); %find crosscorrelation
subplot(212); plot(lags*Ts,c);
%find the lag for maximum absolute crosscorrelation:
L = lags(find(abs(c)==max(abs(c)))); 
estimatedTimeDelay = L*Ts

File: PythonCodeSnippets/snip_signals_time_delay.py
import matplotlib.pyplot as plt
import numpy as np
import pylab as pl
import scipy.signal as sp
from numpy import random as rd

Fs = 8000  # sampling frequency
Ts = 1 / Fs  # Sampling Period
N = 1.5 * Fs  # Define a time of 1.5 second
t = np.arange(0, N) * Ts
print(len(t))
if 1:
    X = rd.rand(int(N)) - 0.5
else:
    X = np.cos(2 * np.pi * 100 * t)
dalayInSamples = 2000
timeDelay = dalayInSamples * Ts
zero = np.zeros(dalayInSamples)
auxX = X[0 : len(X) - dalayInSamples]
y = np.concatenate((zero, auxX))
SNRdb = 10
signalPower = np.mean(X**2)
noisePower = signalPower / (10 ** (SNRdb / 10))
noise = np.sqrt(noisePower) * rd.randn(np.size(y))
y = y + noise
plt.figure()
plt.subplot(211)
plt.plot(t[0:5000], X[0:5000])
plt.plot(t[0:5000], y[0:5000])

plt.subplot(212)

c = sp.correlate(X, y, "full")
lags = sp.correlation_lags(len(X), len(y), "full")

plt.plot(lags * Ts, c)
pl.show()
index_max = max(range(len(c)), key=np.abs(c).__getitem__)
L = lags[index_max]
estimated_time_delay = L * Ts
print(estimated_time_delay)
